File: WanderingSalesman/Episode_5/mcherm/java/MchermEpisode5.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.FileInputStream;
import java.util.HashMap;
import java.util.Map;
import java.math.BigInteger;

/**
 * Code for Episode 4 of the Wandering Salesman problem.
 *
 * Transform it to be iterative instead of recursive.
 */
public class MchermEpisode5 {

    private static JsonNode mapData;
    private static Map<String, Integer> nodeToPosition;


    private static class Matrix {
        public final int numNodes;
        public final BigInteger[][] data;

        /**
         * Make it pretty for output.
         */
        public String toString() {
            final StringBuilder sb = new StringBuilder();
            sb.append("[\n");
            for (int i = 0; i < numNodes; i++) {
                sb.append("  [");
                for (int j = 0; j < numNodes; j++) {
                    sb.append(data[i][j]);
                    sb.append(", ");
                }
                sb.append("],\n");
            }
            sb.append("]");
            return sb.toString();
        }

        /**
         * Create a numNodes x numNodes Matrix populated entirely with
         * value.
         *
         * @param numNodes the side size of this square matrix
         * @param value the single value (or null) which will be used to populate all
         *              cells of the matrix
         */
        public Matrix(final int numNodes, final BigInteger value) {
            this.numNodes = numNodes;
            data = new BigInteger[numNodes][];
            for (int i = 0; i < numNodes; i++) {
                data[i] = new BigInteger[numNodes];
                for (int j = 0; j < numNodes; j++) {
                    data[i][j] = value;
                }
            }
        }

        /**
         * Performs a matrix multiply of this with other and returns the new resulting
         * matrix.
         */
        public Matrix matrixMultiply(final Matrix other) {
            assert other.numNodes == this.numNodes;
            final Matrix result = new Matrix(numNodes, null);
            for (int i = 0; i < numNodes; i++) {
                for (int j = 0; j < numNodes; j++) {
                    BigInteger value = BigInteger.ZERO;
                    for (int k = 0; k < numNodes; k++) {
                        value = value.add(this.data[i][k].multiply(other.data[k][j]));
                    }
                    result.data[i][j] = value;
                }
            }
            return result;
        }

        /**
         * Returns the sum of items in the given row.
         */
        public BigInteger sumOfRow(final int row) {
            BigInteger result = BigInteger.ZERO;
            for (int i = 0; i < numNodes; i++) {
                result = result.add(data[row][i]);
            }
            return result;
        }

        /**
         * Calculates this matrix to a power, using binary exponentiation.
         *
         * DESIGN: We will express power in binary, then repeatedly square this matrix to get
         * powers of two. We'll add together the powers of two that are 1s in the binary
         * number and that gives our answer.
         */
        public Matrix matrixPower(final int power) {
            // --- Create an identity matrix ---
            Matrix result = new Matrix(numNodes, BigInteger.ZERO);
            for (int i = 0; i < numNodes; i++) {
                result.data[i][i] = BigInteger.ONE;
            }

            // --- Set up to calculate <this> to various powers of two by repeated squaring ---
            int powerOfTwo = 1;
            Matrix matrixToPowerOfTwo = this;

            // --- Determine the binary digits of power ---
            final String binaryDigits = Integer.toBinaryString(power);

            // --- Loop through the binary digits ---
            while (true) { // will exit the loop when powerOfTwo reaches the length of the string
                final char binaryDigit = binaryDigits.charAt(binaryDigits.length() - powerOfTwo);
                if (binaryDigit == '1') {
                    result = result.matrixMultiply(matrixToPowerOfTwo);
                }
                powerOfTwo += 1;
                if (powerOfTwo > binaryDigits.length()) {
                    break; // break here instead of in the loop condition to avoid doing an extra matrix multiply
                }
                matrixToPowerOfTwo = matrixToPowerOfTwo.matrixMultiply(matrixToPowerOfTwo);
            }

            // --- Return the result ---
            return result;
        }
    }


    private static BigInteger countPaths(String startNode, int pathLength) {
        // --- This algorithm only works on paths of length 2 and up, so hard-code length 1 ---
        if (pathLength == 1) {
            return BigInteger.ONE;
        }

        // --- Read in the initial matrix ---
        final int numNodes = mapData.get("nodes").size();
        final Matrix adjMatrix = new Matrix(numNodes, null);
        for (int i = 0; i < numNodes; i++) {
            for (int j = 0; j < numNodes; j++) {
                final String value = mapData.get("adjacencyMatrix").get(i).get(j).asText();
                adjMatrix.data[i][j] = new BigInteger(value);
            }
        }

        // --- Perform exponentiation ---
        // we go to the power of pathLength - 1 because doing sumOfRow below gives us one extra level.
        final Matrix outputMatrix = adjMatrix.matrixPower(pathLength - 1);

        // --- Return the answer for this start node ---
        return outputMatrix.sumOfRow(nodeToPosition.get(startNode));
    }


    public static void main(String[] args) throws Exception {
        // --- Read the map data ---
        final FileInputStream in = new FileInputStream("./WanderingSalesman/inputdata/small-map.json");
        mapData = new ObjectMapper().readTree(in);

        // --- Initialize nodeToPosition ---
        nodeToPosition = new HashMap<>();
        int pos = 0;
        for (final JsonNode jsonNode : mapData.get("nodes")) {
            nodeToPosition.put(jsonNode.asText(), pos);
            pos++;
        }

        // --- call the function and print results ---
        int pathLength = 1000000;
        final long startTime = System.currentTimeMillis();
        BigInteger pathCount = countPaths("A", pathLength);
        final long endTime = System.currentTimeMillis();
        System.out.println("Considering paths of length " + pathLength + ", there are " + pathCount + " of them.");
        final long seconds = (endTime - startTime) / 1000;
        System.out.println("Taking " + seconds + " seconds.");
    }
}

File: WanderingSalesman/Episode_5/mcherm/rust/mcherm_episode_5.rs
use std::fs;
use serde::Deserialize;
use serde_json;
use rug::Integer;
use std::time::Instant;


#[allow(non_snake_case)]
#[derive(Deserialize)]
struct MapDescription {
    nodes: Vec<String>,
    adjacencyMatrix: Vec<Vec<u8>>,
}


fn load_map_description() -> Result<MapDescription, Box<dyn std::error::Error>> {
    let s = fs::read_to_string("inputdata/small-map.json")?;
    let answer = serde_json::from_str(&s)?;
    return Ok(answer);
}




type Matrix = Vec<Vec<Integer>>;
type Node<'a> = &'a str;

/// Performs a matrix multiply of m1 and m2.
fn matrix_multiply(m1: &Matrix, m2: &Matrix) -> Matrix {
    assert!(m1.len() == m1[0].len() && m1.len() == m2.len());
    let size = m1.len();
    let mut answer = Matrix::new();

    for i in 0..size {
        let mut row = Vec::with_capacity(size);
        for j in 0..size {
            let mut sum = Integer::ZERO;
            for k in 0..size {
                sum += &m1[i][k] * &m2[k][j];
            }
            row.push(sum);
        }
        answer.push(row);
    }

    // --- Return the result ---
    answer
}


/// Puts a given matrix to a power (efficiently).
fn matrix_power(m: &Matrix, power: u32) -> Matrix {
    // --- Track input matrix to powers of two ---
    let mut power_of_two = m.clone();

    // --- Start with an identity matrix (for length-1 paths you just sit on start_position) ---
    let mut output_matrix = identity_matrix(m.len());

    // --- Go through powers of 2, multiplying them in if the digit is a 1 ---
    let binary_digits = format!("{:b}", power);
    for binary_digit in binary_digits.chars().rev() {
        if binary_digit == '1' {
            output_matrix = matrix_multiply(&output_matrix, &power_of_two);
        }

        power_of_two = matrix_multiply(&power_of_two, &power_of_two);
    }

    // --- Return it ---
    output_matrix
}


/// Creates and returns an identity matrix (1's along the diagonal) of the given size.
fn identity_matrix(size: usize) -> Matrix {
    let mut answer = Matrix::new();
    for i in 0..size {
        let mut row = Vec::with_capacity(size);
        for j in 0..size {
            row.push(if i == j {Integer::from(1)} else {Integer::ZERO});
        }
        answer.push(row);
    }

    // --- Return the result ---
    answer
}


impl MapDescription {
    fn get_adj_matrix(&self) -> Matrix {
        let size = self.adjacencyMatrix.len();
        let mut answer = Vec::with_capacity(size);
        for row in self.adjacencyMatrix.iter() {
            let mut answer_row: Vec<Integer> = Vec::with_capacity(size);
            for item in row {
                answer_row.push(Integer::from(*item));
            }
            answer.push(answer_row);
        }
        answer
    }

    fn node_to_num(&self, node: Node) -> Option<usize> {
        self.nodes.iter().position(|x| x == node)
    }
}


/// Given a map and a start position and number of steps, this returns the count
/// of paths of that length starting from that location.
fn count_paths<'a>(map: &'a MapDescription, start_position: Node<'a>, num_steps: u32) -> Integer {
    let adj_matrix = map.get_adj_matrix();

    // --- Put the matrix to a power ---
    let output_matrix = matrix_power(&adj_matrix, num_steps - 1);

    // --- Sum up the correct row and return that ---
    output_matrix[map.node_to_num(start_position).unwrap()].iter().sum()
}



fn main() -> Result<(), Box<dyn std::error::Error>> {
    let map = load_map_description()?;

    let path_len = 2500000;
    let start_time = Instant::now();
    let path_count = count_paths(&map, &"A", path_len);
    let execute_seconds = start_time.elapsed().as_secs();
    println!("For length {} there are {} paths.", path_len, path_count);
    println!("Taking {} seconds.", execute_seconds);
    return Ok(());
}
